### PR TITLE
Migrate from V1 to V2  Guide

### PR DESCRIPTION
**Pull request by alfred2g**
Migration guide to help customer move from the V1 sdk to the V2 sdk

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by tkc4106 — COMMENTED** [review]
Double check on the structure and references/links to correct samples (for the language).
Add TOC to the top for quick references.
Add code sample folding to simplify the doc and to improve readability.
 Consistency of Connection Type table.

File: documents/MIGRATION_GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,1081 @@
...
...1_unsubscribe_packet.html) object with the help of [UnsubscribePacketBuilder](https://awslabs.github.io/aws-crt-java/software/amazon/awssdk/crt/mqtt5/packets/UnsubscribePacket.UnsubscribePacketBuilder.html). The [Unsubscribe](https://aws.github.io/aws-iot-device-sdk-cpp-v2/class_aws_1_1_crt_1_1_mqtt5_1_1_mqtt5_client.html#a96a931b49893d54712062722c5ab7d1a) operation takes a description of the [UnsubscribePacket](https://aws.github.io/aws-iot-device-sdk-cpp-v2/class_aws_1_1_crt_1_1_mqtt5_1_1_unsubscribe_packet.html) you wish to send and returns a promise that resolves successfully with the corresponding [UnsubAckPacket](https://aws.github.io/aws-iot-device-sdk-cpp-v2/class_aws_1_1_crt_1_1_mqtt5_1_1_un_sub_ack_packet.html) returned by the broker; the promise is rejected with an error if anything goes wrong before the [UnsubAckPacket](https://aws.github.io/aws-iot-device-sdk-cpp-v2/class_aws_1_1_crt_1_1_mqtt5_1_1_un_sub_ack_packet.html) is received. You should always check the reason codes of a [UnsubAckPacket](https://aws.github.io/aws-iot-device-sdk-cpp-v2/class_aws_1_1_crt_1_1_mqtt5_1_1_un_sub_ack_packet.html) completion to determine if the unsubscribe operation actually succeeded.
```
Comment (sfod) [inline documents/MIGRATION_GUIDE.md:339]: One of the links leads to java docs. Also, there are other places in the guide with Java links, I won't mark everyone of them.

File: documents/MIGRATION_GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,1081 @@
...
+
+In V1 SDK, the [awsiotsdk::MqttClient](http://aws-iot-device-sdk-cpp-docs.s3-website-us-east-1.amazonaws.com/latest/classawsiotsdk_1_1_mqtt_client.html) class represents an MQTT client. You instantiate the client directly passing all the required parameters to the class constructor. It’s possible to change client settings after its creation using `set*` methods, e.g. `SetAutoReconnectEnabled` or `SetMaxReconnectBackoffTimeout`.
+
+In V2 SDK, the [Aws::Iot::MqttClient](https://aws.github.io/aws-iot-device-sdk-cpp-v2/class_aws_1_1_iot_1_1_mqtt_client.html) class represents an MQTT client, specifically MQTT5 protocol. V2 SDK provides an [Aws::Iot::Mqtt5ClientBuilder](https://aws.github.io/aws-iot-device-sdk-cpp-v2/class_aws_1_1_iot_1_1_mqtt5_client_builder.html) designed to easily create common configuration types such as direct MQTT or WebSocket connections.
+Once an MQTT5 client is built and finalized, the resulting MQTT5 client cannot have its settings modified.
+
+_Example of creating a client in V1_
+
+```
+#include "OpenSSLConnection.hpp"
```
Comment (sfod) [inline documents/MIGRATION_GUIDE.md:40]: Code blocks should be marked as CPP code. It seems, all code blocks should be fixed.

File: documents/MIGRATION_GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,1081 @@
...
+|	|	|	|	|	|
+|MQTT over Secure WebSocket with AWS SigV4 authentication	|✔	|✔	|	|	|
+|MQTT (over TLS 1.2) with X.509 certificate based mutual authentication	|✘	|✔	|	|X.509	|
+|MQTT with PKCS12 Method	|✔*	|✔	|	|Container for X.509	|
+|MQTT with Custom Key Operation Method	|✔*	|✔	|	|X.509	|
+|MQTT with Custom Authorizer Method	|✔*	|✔	|	|	|
+|MQTT with Windows Certificate Store Method	|✔*	|✔	|	|X.509	|
+|MQTT with PKCS11 Method	|✘	|✔	|	|X.509 plus other formats	|
+|Websocket Connection with Cognito Authentication Method	|✘	|✔	|	|	|
+|HTTP Proxy	|✔**	|✔	|	|	|
```
Comment (sfod) [inline documents/MIGRATION_GUIDE.md:103]: Debatable: Use colors for checks and crosses. I used latex for that: `$${\Large\color{green}&#10004}$$` and `$${\Large\color{orange}&#10004\*}$$`. See in the [java pr](https://github.com/aws/aws-iot-device-sdk-java-v2/pull/533/files#diff-bcb919f53680494fc23f06bec1126f2c8b9ac21d8b01e56b06d77f9fbb31b4acR175).

File: documents/MIGRATION_GUIDE.md [hunk 4]
```diff
@@ -0,0 +1,1081 @@
...
+V1 SDK supports two types of connections to connect to the AWS IoT service: MQTT with X.509 certificate and MQTT over Secure WebSocket with SigV4 authentication.
+
+V2 SDK adds a collection of connection types and cryptography formats (e.g. [PKCS #11](https://en.wikipedia.org/wiki/PKCS_11) and [Custom Authorizer](https://docs.aws.amazon.com/iot/latest/developerguide/custom-authentication.html)), credential providers (e.g. [Amazon Cognito](https://aws.amazon.com/cognito/) and [Windows Certificate Store](https://learn.microsoft.com/en-us/windows-hardware/drivers/install/certificate-stores)), and other connection-related features.
+Refer to the “[Connecting To AWS IoT Core](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/main/documents/MQTT5_Userguide.md#connecting-to-aws-iot-core)” section of the MQTT5 user guide for detailed information and code snippets on each connection type and connection feature.
+
+|Connection Type/Feature	|V1 SDK	|V2 SDK	|User guide section	|	|
+|---	|---	|---	|---	|---	|
+|	|	|	|	|	|
+|MQTT over Secure WebSocket with AWS SigV4 authentication	|✔	|✔	|	|	|
+|MQTT (over TLS 1.2) with X.509 certificate based mutual authentication	|✘	|✔	|	|X.509	|
```
Comment (sfod) [inline documents/MIGRATION_GUIDE.md:96]: The last column in this table is not needed

File: documents/MIGRATION_GUIDE.md [hunk 5]
```diff
@@ -0,0 +1,1081 @@
...
+
+V1 SDK supports two types of connections to connect to the AWS IoT service: MQTT with X.509 certificate and MQTT over Secure WebSocket with SigV4 authentication.
+
+V2 SDK adds a collection of connection types and cryptography formats (e.g. [PKCS #11](https://en.wikipedia.org/wiki/PKCS_11) and [Custom Authorizer](https://docs.aws.amazon.com/iot/latest/developerguide/custom-authentication.html)), credential providers (e.g. [Amazon Cognito](https://aws.amazon.com/cognito/) and [Windows Certificate Store](https://learn.microsoft.com/en-us/windows-hardware/drivers/install/certificate-stores)), and other connection-related features.
+Refer to the “[Connecting To AWS IoT Core](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/main/documents/MQTT5_Userguide.md#connecting-to-aws-iot-core)” section of the MQTT5 user guide for detailed information and code snippets on each connection type and connection feature.
+
+|Connection Type/Feature	|V1 SDK	|V2 SDK	|User guide section	|	|
+|---	|---	|---	|---	|---	|
+|	|	|	|	|	|
+|MQTT over Secure WebSocket with AWS SigV4 authentication	|✔	|✔	|	|	|
```
Comment (sfod) [inline documents/MIGRATION_GUIDE.md:95]: There are no links in the "User guide section" column.

File: documents/MIGRATION_GUIDE.md [hunk 6]
```diff
@@ -0,0 +1,1081 @@
...
+|MQTT with PKCS12 Method	|✔*	|✔	|	|Container for X.509	|
+|MQTT with Custom Key Operation Method	|✔*	|✔	|	|X.509	|
+|MQTT with Custom Authorizer Method	|✔*	|✔	|	|	|
+|MQTT with Windows Certificate Store Method	|✔*	|✔	|	|X.509	|
+|MQTT with PKCS11 Method	|✘	|✔	|	|X.509 plus other formats	|
+|Websocket Connection with Cognito Authentication Method	|✘	|✔	|	|	|
+|HTTP Proxy	|✔**	|✔	|	|	|
+
+✔*   - In order to get this connection type work in V1 SDK, you need to implement the [Custom Authentication workflow](https://docs.aws.amazon.com/iot/latest/developerguide/custom-authorizer.html).
+✔**  - Though V1 does not allow to specify HTTP proxy, it is possible to configure systemwide proxy.
```
Comment (sfod) [inline documents/MIGRATION_GUIDE.md:106]: These two lines are rendered as one line. You should add two spaces or `\` at the end of the first line. I prefer backslash, but it's a non-standard way.

File: documents/MIGRATION_GUIDE.md [hunk 7]
```diff
@@ -0,0 +1,1081 @@
+# Migrate from V1 to V2 of the AWS IoT SDK for C++
+
+The V2 AWS IoT SDK for C++ is a major rewrite of the V1 code base. It includes many updates, such as improved consistency, ease of use, more detailed information about client status, an offline operation queue control, etc. This guide describes the major features that are new in V2, and provides guidance on how to migrate your code to V2 from V1.
+
+## What’s new in V2 SDK
+
+* V2 SDK client is truly async. Operations take callback functions/lambdas, that is called when the operation is registered with the server
+    Blocking calls can be emulated by setting an `std::promise<>` in the callback by calling `promise.set_value() `and then waiting for the returned `std::future<>` object to be resolved after the function call by calling `future.wait()`
+* V2 SDK provides implementation for MQTT5 protocol, the next step in evolution of the MQTT protocol.
+* Public API terminology has changed. You `S``tart()` or `S``top()` the MQTT5 client rather than `Connect` or `Disconnect` like in V1. This removes the semantic confusion with the connect/disconnect as the client-level controls vs. internal recurrent networking events.
```
Comment (sfod) [inline documents/MIGRATION_GUIDE.md:10]: Extra "`" symbol in Start and Stop words.

File: documents/MIGRATION_GUIDE.md [hunk 8]
```diff
@@ -0,0 +1,1081 @@
...
+```
+
+In V2 logging is shutdown automatically with ApiHandle destruction when it goes out of scope
+
+### Client for Device Shadow Service
+
+V1 SDK is built with [AWS IoT device shadow support](http://docs.aws.amazon.com/iot/latest/developerguide/iot-thing-shadows.html), providing access to thing shadows (sometimes referred to as device shadows). It also supports a simplified shadow access model, which allows developers to exchange data with their shadows by just using getter and setter methods without having to serialize or deserialize any JSON documents.
+
+V2 SDK supports device shadow service as well, but with completely different API.
+First, you subscribe to special topics to get data and feedback from a service. The service client provides API for that. For example, `SubscribeToGetShadow``Accepted`  subscribes to a topic to which AWS IoT Core will publish a shadow document; and via the `SubscribeToGetShadowRejected` the server will notify you if it cannot send you a requested document.
```
Comment (sfod) [inline documents/MIGRATION_GUIDE.md:572]: Extra "\`" symbols in "`SubscribeToGetShadow``Accepted`"

File: documents/MIGRATION_GUIDE.md [hunk 9]
```diff
@@ -0,0 +1,1081 @@
...
+
+```
+builder->WithClientConnectionSuccessCallback(
+        [&](const Mqtt5::OnConnectionSuccessEventData &eventData) { });
+
+builder->WithClientConnectionFailureCallback([&](
+        const Mqtt5::OnConnectionFailureEventData &eventData) { });
+
+builder->WithClientStoppedCallback([&](
+        const Mqtt5::OnStoppedEventData & eventData) { });
```
Comment (sfod) [inline documents/MIGRATION_GUIDE.md:181]: Trivial: extra space after `&`

**Review by sfod — CHANGES_REQUESTED** [review]

File: documents/MIGRATION_GUIDE.md (hunk 9 above)
Comment (tkc4106) [reply documents/MIGRATION_GUIDE.md:181]: My comments above were meant as To-dos. I spoke to alfred2g@ about them earlier. I update the comment with "to-do" to clarify.

File: documents/MIGRATION_GUIDE.md [hunk 10]
```diff
@@ -0,0 +1,1219 @@
...
+| MQTT with Windows Certificate Store Method               | $${\Large\color{orange}&#10004\*}$$     | $${\Large\color{green}&#10004}$$ | [link](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/main/documents/MQTT5_Userguide.md#direct-mqtt-with-windows-certificate-store-method) |
+| MQTT with PKCS11 Method                                  | $${\Large\color{red}&#10008}$$          | $${\Large\color{green}&#10004}$$ | [link](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/main/documents/MQTT5_Userguide.md#direct-mqtt-with-pkcs11-method) |
+| HTTP Proxy                                               | $${\Large\color{orange}&#10004\*\*}$$   | $${\Large\color{green}&#10004}$$ | [link](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/main/documents/MQTT5_Userguide.md#adding-an-http-proxy) |
+
+${\Large\color{orange}&#10004\*}$ - To get this connection type work in the v1 SDK, you need to implement the [Custom Authentication workflow](https://docs.aws.amazon.com/iot/latest/developerguide/custom-authorizer.html).\
+${\Large\color{orange}&#10004\*\*}$ - The  v1 SDK does not allow to specify HTTP proxy, but sustemwide proxy
```
Comment (alfred2g) [inline documents/MIGRATION_GUIDE.md:151]: typo: systemwide

**Review by sfod — APPROVED** [review]

File: documents/MIGRATION_GUIDE.md [hunk 11]
```diff
@@ -0,0 +1,1451 @@
...
+
+> [!NOTE]
+> If you can't find the information you need in this guide, visit the [How to Get Help](#how-to-get-help) section for
+> more help and guidance
+
+* [What's new in AWS IoT Device SDK for C++ v2](#whats-new-in-aws-iot-device-sdk-for-c-v2)
+* [How to get started with the SDK for C++ v2](#how-to-get-started-with-the-sdk-for-c-v2)
+    * [Mqtt protocol](#mqtt-protocol)
+    * [Client builder](#client-builder)
+    * [Client Start](#client-start)
```
[the remark this answers is not in the document]
Comment (alfred2g) [reply documents/MIGRATION_GUIDE.md:16]: it is capital in C++

**Pull request merged**